### PR TITLE
[SPARK-7501] [Streaming] DAG visualization: show DStream operations

### PR DESCRIPTION
**Pull request by andrewor14**
This is similar to #5999, but for streaming. Roughly 200 lines are tests.

One thing to note here is that we already do some kind of scoping thing for call sites, so this patch adds the new RDD operation scoping logic in the same place. Also, this patch adds a `try finally` block to set the relevant variables in a safer way.

@tdas @zsxwing

---

**Before**

---

**After**

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -1523,13 +1523,15 @@ abstract class RDD[T: ClassTag](
...
-        dependencies.foreach(_.rdd.doCheckpoint())
+  private[spark] def doCheckpoint(): Unit = {
+    RDDOperationScope.withScope(sc, "checkpoint", false, true) {
+      if (!doCheckpointCalled) {
+        doCheckpointCalled = true
+        if (checkpointData.isDefined) {
+          checkpointData.get.doCheckpoint()
+        } else {
+          dependencies.foreach(_.rdd.doCheckpoint())
+        }
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: This is not for this PR. The other PR still not merged in master .. isnt it?

File: core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala [hunk 2]
```diff
@@ -43,7 +43,7 @@ import org.apache.spark.SparkContext
 @JsonPropertyOrder(Array("id", "name", "parent"))
 private[spark] class RDDOperationScope(
     val name: String,
-    val parent: Option[RDDOperationScope] = None) {
+    val parent: Option[RDDOperationScope] = None) extends Serializable {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala]: Why was this made serializable?

**Comment by tdas** [comment]
I think there should be a few unit tests to test this. See how the creation sites of the DStream are tested in StreamingContextSuite, test "DStream and generated RDD creation sites"

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 3]
```diff
@@ -109,6 +109,18 @@ abstract class DStream[T: ClassTag] (
...
 
+  /**
+   * The scope associated with the operation that created this DStream.
+   *
+   * This is the medium through which we pass the DStream operation name (e.g. updatedStateByKey)
+   * to the RDDs created by this DStream. Note that we never directly use this scope in RDDs.
+   * Instead, every time we call `compute` we instantiate a new scope using the same name as this
+   * one. Otherwise, all RDDs ever created by this DStream will be in the same scope.
+   */
+  private val scope: Option[RDDOperationScope] = {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Can you store only the name in as a DStream field? I does not make sense to save this arbitrary mostly unnecessary RDDOperationScope object to be serialized with DStreams as part of DStream checkpoints.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 4]
```diff
@@ -321,6 +320,48 @@ abstract class DStream[T: ClassTag] (
   }
 
   /**
+   * Helper method to generate an RDD for the specified time.
+   * This sets and resets the relevant local variables before and after the call to compute.
+   */
+  private def doCompute(time: Time): Option[RDD[T]] = {
+    val scopeKey = SparkContext.RDD_SCOPE_KEY
+    val scopeNoOverrideKey = SparkContext.RDD_SCOPE_NO_OVERRIDE_KEY
+    // Set the thread-local property for call sites to this DStream's creation site
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Can you update the comments to generalize it to both call sites and scopes?

**Comment by tdas** [comment]
This doesnt yet set the scope for input stream (that is DStream that are directly created, not using DStream transformations). I think you need to use the withScope in places like StreamingContext.socketStream(). Example: 

```
class StreamingContext {

def socketTextStream(): DStream[String] = withScope {
   new SocketInputDStream[String](...)
}
```

This needs to be done for every subclass of `InputDStream` and `ReceiverInputDStream`

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: oops I can't believe I accidentally merged these two PRs.

**Comment by andrewor14** [comment]
@zsxwing @tdas Check out the latest screenshot!! Also cc/ @JoshRosen who might be interested.

**Comment by andrewor14** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala [hunk 5]
```diff
@@ -43,10 +43,9 @@ import org.apache.spark.SparkContext
 @JsonPropertyOrder(Array("id", "name", "parent"))
 private[spark] class RDDOperationScope(
     val name: String,
+    val id: String = RDDOperationScope.nextScopeId().toString,
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala]: Super Nit: typically default params are ordered based on what is more likely to get explicitly specified. I think `parent` is more likely be specified by callers than the `id`, so should be ordered accordingly.

File: core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala [hunk 6]
```diff
@@ -88,14 +87,25 @@ private[spark] object RDDOperationScope {
 
   /**
    * Execute the given body such that all RDDs created in this body will have the same scope.
-   * The name of the scope will be the name of the method that immediately encloses this one.
+   * The name of the scope will be the first method name in the stack trace that is not the
+   * same as this method's.
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala]: What do you mean by `this method's `? The name is not `withScope`?

File: core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala [hunk 7]
```diff
@@ -88,14 +87,25 @@ private[spark] object RDDOperationScope {
...
    *
    * Note: Return statements are NOT allowed in body.
    */
   private[spark] def withScope[T](
       sc: SparkContext,
       allowNesting: Boolean = false)(body: => T): T = {
-    val callerMethodName = Thread.currentThread.getStackTrace()(3).getMethodName
+    val stackTrace = Thread.currentThread.getStackTrace().tail // ignore "Thread#getStackTrace"
+    val ourMethodName = stackTrace(1).getMethodName
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala]: Which method name is this? `withScope`? If so mention that in the comments.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 8]
```diff
@@ -338,6 +370,48 @@ abstract class DStream[T: ClassTag] (
...
+   * This sets and resets the relevant local variables before and after the call to compute.
+   */
+  private def doCompute(time: Time): Option[RDD[T]] = {
+    val scopeKey = SparkContext.RDD_SCOPE_KEY
+    val scopeNoOverrideKey = SparkContext.RDD_SCOPE_NO_OVERRIDE_KEY
+    // Pass this DStream's operation scope and creation site information to RDDs through
+    // thread-local properties in our SparkContext. Since this method may be called from another
+    // DStream, we need to temporarily store any old scope and creation site information to
+    // restore them later after setting our own.
+    // TODO: this won't work if multiple StreamingContexts share the same SparkContext
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Why wont this work? Each streaming context will be calling compute in its own scheduling thread. They are not going to interfere with each other.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala [hunk 9]
```diff
@@ -49,6 +49,9 @@ abstract class InputDStream[T: ClassTag] (@transient ssc_ : StreamingContext)
    */
   private[streaming] def name: String = s"${getClass.getSimpleName}-$id"
 
+  /** Human-friendly scope name to use in place of generic operation names (e.g. createStream). */
+  protected override val customScopeName: Option[String] = Some(s"input stream [$id]")
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: Why `protected` and not `private[streaming]` like `name` ?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 10]
```diff
@@ -126,6 +126,8 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
   @transient private var path_ : Path = null
   @transient private var fs_ : FileSystem = null
 
+  protected override val customScopeName: Option[String] = Some(s"file stream [$id]")
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Technically, this should/could be either `file stream` and `text file stream` depending on whether `ssc.fileStream` or `ssc.textFileStream` created it.  But I am okay as is.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 11]
```diff
@@ -111,6 +112,50 @@ abstract class DStream[T: ClassTag] (
...
+   */
+  private[streaming] val baseScope: Option[String] = {
+    Option(ssc.sc.getLocalProperty(SparkContext.RDD_SCOPE_KEY))
+  }
+
+  /**
+   * An optional custom name for all scopes generated by this DStream.
+   * If None, the name of the operation that created this DStream will be used.
+   */
+  protected val customScopeName: Option[String] = None
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Lets not expose this to the public API. Rather do `private[streaming]` like `createSite` and `baseScope`

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 12]
```diff
@@ -302,35 +347,22 @@ abstract class DStream[T: ClassTag] (
...
-        ssc.sparkContext.setCallSite(creationSite)
-        // Disable checks for existing output directories in jobs launched by the streaming
-        // scheduler, since we may need to write output to an existing directory during checkpoint
-        // recovery; see SPARK-4835 for more details. We need to have this call here because
-        // compute() might cause Spark jobs to be launched.
-        val rddOption = PairRDDFunctions.disableOutputSpecValidation.withValue(true) {
-          compute(time)
-        }
-        ssc.sparkContext.setCallSite(prevCallSite)
+        val newRDD = doCompute(time)
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: This should have `option` in the name as before.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala (hunk 11 above)
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Wait, but protected is actually tighter than `private[streaming]`. This means only subclasses can access / override it

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala (hunk 9 above)
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: because this is only used in `DStream` and overridden in its children. Classes outside of `DStream` don't need to access it.

File: core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala (hunk 7 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala]: I didn't do it cause if we change the method name then we must update the comment. It's a little like duplicating code. But I can add it not a big deal

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala (hunk 8 above)
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: I see. OK.

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala [hunk 13]
```diff
@@ -40,7 +40,7 @@ object TwitterUtils {
       twitterAuth: Option[Authorization],
       filters: Seq[String] = Nil,
       storageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK_SER_2
-    ): ReceiverInputDStream[Status] = {
+    ): ReceiverInputDStream[Status] = ssc.withScope {
```
Comment (tdas) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: Why have this, when you are any ways overriding it with `customScopeName`?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala (hunk 11 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: But we dont want to expose it publicly. That's why may be its best to do `protected[streaming]`.

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala (hunk 13 above)
Comment (tdas) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: Wait, why at all have customScopeName in DStream, when you can always set the custom scope name using `withScope(someScopeName)`?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala (hunk 11 above)
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Do we not want users who create their own `DStreams` to be able to give their own names?

Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Ok, `protected[streaming]` it is

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala (hunk 13 above)
Comment (andrewor14) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: Answer to your first question: this is still need because you need the scope ID even if you don't need the name.

2nd question: It's to avoid duplicating the name of the stream. There are a few methods that create `*KafkaInputDStream` for instance.

Comment (tdas) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: Well you are sort-of-duplicating some code/processing by adding `withScope` as well as `customScopeName`. Instead just add `withScope("kafka stream")` in one place and remove the extra `customeScopeName` from everywhere. And even though there are many versions of KafkaUtils.createStream, only one or two of them actually calls `new KafkaInputDStream`. So using `withScope("kafka stream")` in only those is sufficient. Also that more matches other ways of setting scope where the scope is defined by the caller method rather than by the entity itself. In fact, conceptually, an entity cannot really define the idea of scope around itself; only the caller/user of the entity can define the scope around the entity.

Comment (andrewor14) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: then how do we include the DStream IDs?

Comment (andrewor14) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: Alright, I implemented what you suggested in the latest commit. I had to add another helper method that `InputDStream` overrides to include the ID. For most streams there is not much duplication, but for `FileInputDStream`, for instance, we have to put `"file stream"` in 3 places. I think there are dis/advantages in both approaches. Let me know what you think about the latest one.

File: external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala [hunk 14]
```diff
@@ -37,7 +37,7 @@ object MQTTUtils {
       brokerUrl: String,
       topic: String,
       storageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK_SER_2
-    ): ReceiverInputDStream[String] = {
+    ): ReceiverInputDStream[String] = ssc.withNamedScope("MQTT stream") {
```
Comment (tdas) [inline external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala]: Make lower case.

**Comment by andrewor14** [comment]
@tdas PTAL. I refactored the scope naming thing based on our offline discussions and fixed the `foreachRDD` case. It's ready for merge from my side.

**Comment by andrewor14** [comment]
retest this please

File: external/flume/src/main/scala/org/apache/spark/streaming/flume/FlumeUtils.scala [hunk 15]
```diff
@@ -60,10 +60,7 @@ object FlumeUtils {
       storageLevel: StorageLevel,
       enableDecompression: Boolean
     ): ReceiverInputDStream[SparkFlumeEvent] = {
-    val inputStream = new FlumeInputDStream[SparkFlumeEvent](
```
Comment (tdas) [inline external/flume/src/main/scala/org/apache/spark/streaming/flume/FlumeUtils.scala]: No need to change this file any more. isnt it?

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala [hunk 16]
```diff
@@ -21,8 +21,8 @@ import twitter4j.Status
 import twitter4j.auth.Authorization
 import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.StreamingContext
-import org.apache.spark.streaming.api.java.{JavaReceiverInputDStream, JavaDStream, JavaStreamingContext}
-import org.apache.spark.streaming.dstream.{ReceiverInputDStream, DStream}
```
Comment (tdas) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala]: Lets not change this file with this patch.

File: external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala [hunk 17]
```diff
@@ -21,8 +21,8 @@ import scala.reflect.ClassTag
 
 import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.StreamingContext
-import org.apache.spark.streaming.api.java.{JavaReceiverInputDStream, JavaStreamingContext, JavaDStream}
-import org.apache.spark.streaming.dstream.{ReceiverInputDStream, DStream}
```
Comment (tdas) [inline external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala]: Lets not change this file with this patch since not needed.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 18]
```diff
@@ -338,6 +369,41 @@ abstract class DStream[T: ClassTag] (
   }
 
   /**
+   * Wrap a body of code such that the call site and operation scope
+   * information are passed to the RDDs created in this body properly.
+   */
+  protected def createRDDWith[U](time: Time)(body: => U): U = {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: the name is incomplete. `createRDDWith` does not make sense. It has to `withSomething`. Maybe `withEnv`?
Also why did you change it from `doCompute`? Is there are point in wrapping whole of `getOrCompute` in this function, rather than just the `compute()` function as you were doing it earlier with `doCompute`?

Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: This method is hard to name, but I don't think `withEnv` is a descriptive name at all. If you can come up with others, I am open to changing it.

I put it in its own method because we need to reuse it in `ForEachDStream`

File: external/flume/src/main/scala/org/apache/spark/streaming/flume/FlumeUtils.scala (hunk 15 above)
Comment (andrewor14) [inline external/flume/src/main/scala/org/apache/spark/streaming/flume/FlumeUtils.scala]: yes, but we're literally putting it in a val here and just returning it...

File: external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala (hunk 17 above)
Comment (andrewor14) [inline external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala]: alright, I will add these unused imports back...

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala (hunk 18 above)
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: As discussed offline, we will rename this to `createRDDWithLocalProperties`

**Comment by tdas** [comment]
LGTM! Thanks @andrewor14 for this patch.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
Great, merging into master 1.4.
